discovery_engine_core/web-api/src/storage.rs

```diff
@@ -44,37 +58,115 @@ impl UserState {
+                last_view: coi.last_view.into(),
+            })
+            .collect();
+
+        Ok(Some(UserInterests { positive, negative }))
     }
 
     pub(crate) async fn update(
         &self,
         id: &UserId,
         user_interests: &UserInterests,
     ) -> Result<(), GenericError> {
-        let serialized_state = bincode::serialize(user_interests)?;
+        let positive_cois = &user_interests.positive;
+        let negative_cois = &user_interests.negative;
 
         let mut tx = self.pool.begin().await?;
-
-        sqlx::query(
-            "INSERT INTO user_state(id, state) VALUES ($1, $2)
-                ON CONFLICT (id) DO UPDATE SET state = EXCLUDED.state;",
-        )
-        .bind(id.as_ref())
-        .bind(serialized_state)
-        .execute(&mut tx)
-        .await?;
+        let mut query_builder = QueryBuilder::new("INSERT INTO ");
+
+        // The amount of rows that we can store via bulk inserts
+        // (<https://docs.rs/sqlx-core/latest/sqlx_core/query_builder/struct.QueryBuilder.html#method.push_values>)
+        // is limited by the postgreSQL bind limit. Hence, the BIND_LIMIT is divided by the number of
+        // fields in the largest tuple.
+        for cois in positive_cois.chunks(BIND_LIMIT / 7) {
+            query_builder
+                .reset()
+                .push("center_of_interest (coi_id, user_id, is_positive, embedding, view_count, view_time_ms, last_view) ")
+                .push_values(cois, |mut stm, coi| {
+                    let timestamp: DateTime<Utc> = coi.stats.last_view.into();
+
+                    // fine as we convert it back to usize/u64 when we fetch it from the database
+                    #[allow(clippy::cast_possible_wrap, clippy::cast_possible_truncation)]
+                    stm.push_bind(coi.id.as_ref())
+                    .push_bind(id.as_ref())
+                    .push_bind(true)
+                    .push_bind(coi.point.to_vec())
+                        .push_bind(coi.stats.view_count as i32)
+                        .push_bind(coi.stats.view_time.as_millis() as i32)
+                        .push_bind(timestamp);
+                })
+                .push(
+                    " ON CONFLICT (coi_id) DO UPDATE SET
+                        embedding = EXCLUDED.embedding,
+                        view_count = EXCLUDED.view_count,
+                        view_time_ms = EXCLUDED.view_time_ms,
+                        last_view = EXCLUDED.last_view;"
+                )
+                .build()
+                .persistent(false)
+                .execute(&mut *tx)
+                .await?;
+        }
+
+        for cois in negative_cois.chunks(BIND_LIMIT / 5) {
+            query_builder
+                .reset()
+                .push("center_of_interest (coi_id, user_id, is_positive, embedding, last_view) ")
+                .push_values(cois, |mut stm, coi| {
+                    let timestamp: DateTime<Utc> = coi.last_view.into();
```

Comment:
:pushpin:  Given that we now anyway go through `DateTime<Utc>` it could make sense to use it as type for `last_view` in the future.